File: king-admin-java/src/main/java/com/oukingtim/service/impl/SysMenuServiceImpl.java
package com.oukingtim.service.impl;

import com.baomidou.mybatisplus.mapper.EntityWrapper;
import com.baomidou.mybatisplus.service.impl.ServiceImpl;
import com.oukingtim.domain.SysMenu;
import com.oukingtim.domain.SysRoleMenu;
import com.oukingtim.mapper.SysMenuMapper;
import com.oukingtim.mapper.SysRoleMenuMapper;
import com.oukingtim.service.SysMenuService;
import com.oukingtim.util.MenuTreeUtil;
import com.oukingtim.web.vm.JsTreeVM;
import com.oukingtim.web.vm.MenuVM;
import com.oukingtim.web.vm.TreeStateVM;
import org.apache.commons.lang.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.*;

/**
 * Created by oukingtim
 */
@Service
public class SysMenuServiceImpl extends ServiceImpl<SysMenuMapper, SysMenu> implements SysMenuService {

    @Autowired
    private SysRoleMenuMapper sysRoleMenuMapper;

    @Override
    public List<JsTreeVM> getMenuTree(String roleId) {
        List<JsTreeVM> treeVoList = new ArrayList<>();
        EntityWrapper<SysMenu> wrapper = new EntityWrapper<SysMenu>();
        wrapper.orderBy("sys_menu.order");
        List<SysMenu> menuList = selectList(wrapper);
        List<String> menuIds = new ArrayList<>();
        if (StringUtils.isNotBlank(roleId)) {
            menuIds = sysRoleMenuMapper.selectMenuIdsByRoleId(roleId);
        }
        for (SysMenu sysMenu : menuList) {
            JsTreeVM jsTreeVM = new JsTreeVM();
            jsTreeVM.setId(sysMenu.getId());
            jsTreeVM.setParent(sysMenu.getParentId()==null?"#":sysMenu.getParentId());
            jsTreeVM.setText(sysMenu.getTitle());
            jsTreeVM.setIcon(sysMenu.getIcon());
            TreeStateVM state = new TreeStateVM();
            if (menuIds.size()>0&&menuIds.contains(sysMenu.getId())){
                state.setSelected(true);
            }
            jsTreeVM.setState(state);
            treeVoList.add(jsTreeVM);
        }
        return treeVoList;
    }

    @Override
    public boolean insert(SysMenu entity) {
        boolean flag = super.insert(entity);

        //自动分配给拥有父菜单的角色权限
        List<SysRoleMenu> roleMenuList = sysRoleMenuMapper.selectList(
                new EntityWrapper<>(new SysRoleMenu(entity.getParentId())));
        for (SysRoleMenu roleMenu: roleMenuList){
            SysRoleMenu sysRoleMenu = new SysRoleMenu();
            sysRoleMenu.setMenuId(entity.getId());
            sysRoleMenu.setRoleId(roleMenu.getRoleId());
            sysRoleMenuMapper.insert(sysRoleMenu);
        }
        return flag;
    }

    @Override
    public Set<String> getPermissions(String userId) {

        List<String> menuList = baseMapper.getPermissions(userId);
        //用户权限列表
        Set<String> permsSet = new HashSet<String>();
        for(String perm : menuList){
            if(StringUtils.isBlank(perm)){
                continue;
            }
            permsSet.addAll(Arrays.asList(perm.trim().split(",")));
        }
        return permsSet;
    }

    @Override
    public List<MenuVM> getMenu(String userId) {
        List<SysMenu> menuList = baseMapper.selectMenuByUserId(userId);
        return MenuTreeUtil.getMenu(menuList);
    }
}

File: king-admin-kotlin/src/main/kotlin/com/oukingtim/service/impl/SysMenuServiceImpl.kt
package com.oukingtim.service.impl

import com.baomidou.mybatisplus.mapper.EntityWrapper
import com.baomidou.mybatisplus.service.impl.ServiceImpl
import com.oukingtim.domain.SysMenu
import com.oukingtim.mapper.SysMenuMapper
import com.oukingtim.mapper.SysRoleMenuMapper
import com.oukingtim.service.SysMenuService
import com.oukingtim.util.MenuTreeUtil
import com.oukingtim.web.vm.JsTreeVM
import com.oukingtim.web.vm.MenuVM
import com.oukingtim.web.vm.TreeStateVM
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.*

/**
 * Created by oukingtim
 */
@Service
class SysMenuServiceImpl : ServiceImpl<SysMenuMapper, SysMenu>(), SysMenuService {

    @Autowired
    private val sysRoleMenuMapper: SysRoleMenuMapper? = null

    override fun getMenuTree(roleId: String): List<JsTreeVM> {
        val treeVoList = ArrayList<JsTreeVM>()
        val wrapper = EntityWrapper<SysMenu>()
        wrapper.orderBy("sys_menu.order")
        val menuList = selectList(wrapper)
        var menuIds: List<String> = ArrayList()
        if (!roleId.isNullOrBlank()) {
            menuIds = sysRoleMenuMapper!!.selectMenuIdsByRoleId(roleId)
        }
        for (sysMenu in menuList) {
            val jsTreeVM = JsTreeVM()
            jsTreeVM.id = sysMenu.id
            jsTreeVM.parent = if (sysMenu.parentId.isNullOrBlank()) "#" else sysMenu.parentId
            jsTreeVM.text = sysMenu.title
            jsTreeVM.icon = sysMenu.icon
            val state = TreeStateVM()
            if (menuIds.size > 0 && menuIds.contains(sysMenu.id)) {
                state.selected = true
            }
            jsTreeVM.state = state
            treeVoList.add(jsTreeVM)
        }
        return treeVoList
    }

    override fun getPermissions(userId: String): Set<String> {
        val menuList = baseMapper.getPermissions(userId)
        //用户权限列表
        val permsSet = HashSet<String>()
        for (perm in menuList) {
            if (perm.isNullOrBlank()) {
                continue
            }
            permsSet.addAll(Arrays.asList(*perm.trim { it <= ' ' }.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()))
        }
        return permsSet
    }

    override fun getMenu(userId: String) = MenuTreeUtil.getMenu(baseMapper.selectMenuByUserId(userId))!!
}